src/common/instructions.rs

```diff
@@ -0,0 +1,187 @@
+            }
+        }
+
+        // Get discriminators for named instructions if any exist
+        let named_instructions: Vec<String> = instruction_names
+            .iter()
+            .filter(|i| !i.starts_with('[') && **i != "*")
+            .cloned()
+            .collect();
+
+        if !named_instructions.is_empty() {
+            let idl_discriminators = get_instruction_discriminators(
+                rpc_client,
+                &config.program_id,
+                &named_instructions,
+            ).await?;
+            
+            // Flip key-value pairs when extending
+            for (name, disc) in idl_discriminators {
+                discriminator_map.insert(disc, name);
+            }
+        }
+
+        configs_with_discriminators.push(ProgramInstructionConfigWithDiscriminators {
+            program_id: config.program_id.clone(),
+            instructions: instruction_names,
+            discriminators: discriminator_map,
+        });
+    }
+
+    Ok(configs_with_discriminators)
+}
+
+
+
+/// Gets the on-chain IDL for a program and matches instruction discriminators
+pub async fn get_instruction_discriminators(
+    rpc_client: &RpcClient,
+    program_id: &str,
+    instruction_names: &[String],
+) -> Result<HashMap<String, Vec<u8>>> {
+    let program_pubkey = Pubkey::from_str(program_id)?;
+    
+    // Find account that contains the IDL
+    // Taken from https://github.com/coral-xyz/anchor/blob/5d0fe35a455bc98d83e03cecdd257dfb91e8f820/lang/syn/src/codegen/program/idl.rs#L17-L26
+    let program_signer = Pubkey::find_program_address(&[], &program_pubkey).0;
+    let idl_address = Pubkey::create_with_seed(&program_signer, ANCHOR_DISCRIMINATOR, &program_pubkey)?;
+
+    // Get IDL account data
+    let idl_account = rpc_client.get_account(&idl_address).await?;
+    
+    // Parse IDL data
+    // TODO: v0.31 and above versions of anchor do not require the discriminator to be 8 bytes...
+    let mut d: &[u8] = &idl_account.data[8..]; // Skip 8-byte discriminator
+    
+    #[derive(BorshDeserialize)]
+    struct IdlAccountData {
+        authority: Pubkey,
+        data_len: u32,
+        data: Vec<u8>,
+    }
+    
+    let idl_account: IdlAccountData = borsh::BorshDeserialize::deserialize(&mut d)?;
+
+    let compressed_len: usize = idl_account.data_len.try_into().unwrap();
+    let compressed_bytes = &idl_account.data[..compressed_len];
+    let mut z = ZlibDecoder::new(compressed_bytes);
+    let mut s = Vec::new();
+    z.read_to_end(&mut s)?;
+    let idl: serde_json::Value = serde_json::from_slice(&s[..])?;
+
+    let mut discriminators = HashMap::new();
+
+    // Get instructions from IDL
+    if let Some(instructions) = idl["instructions"].as_array() {
+        for instruction in instructions {
+            if let Some(name) = instruction["name"].as_str() {
+                // Convert instruction name to snake_case for matching
+                let snake_case = to_snake_case(name);
```

Comment:
Thinking of recommending all provided instructions in the config to be snake case (or I can just automatically convert) so we fuzzy match instead of requiring exact matches